pysnn/learning.py

```diff
@@ -73,39 +77,107 @@ class MSTDPET(LearningRule):
+
+        # Collect desired tensors from state dict in a layer object
+        for idx, layer in enumerate(layers):
+            new_layer = {}
+            new_layer["pre_spikes"] = layer["pre_neuron"]["spikes"]
+            new_layer["pre_trace"] = layer["pre_neuron"]["trace"]
+            new_layer["post_spikes"] = layer["post_neuron"]["spikes"]
+            new_layer["post_trace"] = layer["post_neuron"]["trace"]
+            new_layer["weight"] = layer["connection"]["weight"]
+            new_layer["e_trace"] = torch.zeros_like(layer["connection"]["weight"])
+            layers[idx] = new_layer
+
+        self.a_pre = a_pre
+        self.a_post = a_post
+        self.lr = lr
+        self.e_trace_decay = e_trace_decay
+
+        # To possibly later support groups, without chaning interface
+        defaults = {
+            "a_pre": a_pre,
+            "a_post": a_post,
+            "lr": lr,
+            "e_trace_decay": e_trace_decay,
+        }
+
+        super(MSTDPET, self).__init__(layers, defaults)
+
+    def update_state(self):
         r"""Update eligibility trace based on pre and postsynaptic spiking activity.
         
         This function has to be called manually after each timestep. Should not be called from within forward, 
         as this does is likely not called every timestep.
         """
 
-        self.e_trace *= self.e_trace_decay
-        self.e_trace += self.a_pre * torch.ger(
-            self.post_neuron.spikes.view(-1).float(), self.pre_neuron.trace.view(-1)
-        ) - self.a_post * torch.ger(
-            self.post_neuron.trace.view(-1), self.pre_neuron.spikes.view(-1).float()
-        )
+        for layer in self.layers:
+            # Update eligibility trace
+            layer["e_trace"] *= self.e_trace_decay
+            layer["e_trace"] += self.a_pre * torch.ger(
+                layer["post_spikes"].view(-1).float(), layer["pre_trace"].view(-1)
+            ) - self.a_post * torch.ger(
+                layer["post_trace"].view(-1), layer["pre_spikes"].view(-1).float()
+            )
+
+    def reset_state(self):
+        for layer in self.layers:
+            layer["e_trace"].fill_(0)
 
-    def forward(self, reward):
+    def step(self, reward):
         # TODO: add weight clamping?
-        self.connection.weight += self.lr * reward * self.e_trace
+        for layer in self.layers:
+            layer["weight"] += self.defaults["lr"] * reward * layer["e_trace"]
+
+
+#########################################################
+# Fede STDP
+#########################################################
+class FedeSTDP(LearningRule):
+    r"""STDP version for Paredes Valles, performs mean operation over the batch 
+    dimension before weight update."""
+
+    def __init__(self, layers, lr, w_init, a):
+        # Check layer formats
+        self.check_layers(layers)
+
+        # Set default hyper parameters
+        self.lr = lr
+        self.w_init = w_init
+        self.a = a
```

Comment:
`lr` should be > 0 and `a` should be < 1 for stability (see paper), so maybe check for this? Or at least warn?